% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/MetaSTAARlite_worker_cov_cond.R
\name{MetaSTAARlite_worker_cov_cond}
\alias{MetaSTAARlite_worker_cov_cond}
\title{Generating summary statistics file for conditional analysis using MetaSTAARlite (the "worker" step)}
\usage{
MetaSTAARlite_worker_cov_cond(
  genotype,
  genotype_adj,
  obj_nullmodel,
  variant_info,
  variant_adj_info
)
}
\arguments{
\item{genotype}{an n*p genotype matrix (dosage matrix) of the target sequence,
where n is the sample size and p is the number of genetic variants. If the input genotype matrix
is sparse (e.g. \code{dgCMatrix} format), it is assumed that it has been flipped to represent
minor allele coding.}

\item{genotype_adj}{an n*p_adj genotype matrix (dosage matrix) of the target sequence, where n is
the sample size and p_adj is the number of genetic variants to be adjusted for in conditional
analysis (or a vector of a single variant with length n if p_adj is 1).}

\item{obj_nullmodel}{an object from fitting the null model, which is either the output from \code{\link{fit_nullmodel}} function
in the \code{\link{STAARpipeline}} package, or the output from \code{fitNullModel} function in the \code{GENESIS} package
and transformed using the \code{\link{genesis2staar_nullmodel}} function in the \code{\link{STAARpipeline}} package.}

\item{variant_info}{a data frame or matrix of variant information (unique identifier)
with p rows (listed in the same order as the columns of \code{genotype}) and should contain
the following 4 columns: chromosome (chr), position (pos), reference allele (ref), and alternative allele (alt).}

\item{variant_adj_info}{a data frame or matrix of adjusted variant information (unique identifier)
with p_adj rows (listed in the same order as the rows of \code{genotype_adj}) and should contain
the following 4 columns: chromosome (chr), position (pos), reference allele (ref), and alternative allele (alt).}
}
\value{
a list with the following members:

\code{GTPG_cond}: the covariance matrix between all variants in the variant-set (rows)
and all variants in the conditional variant-set (columns) (the covariance file
for conditional analysis).

\code{variant_info}: the data frame or matrix of variant information (unique identifier)
with p rows (listed in the same order as the rows of \code{GTPG_cond}) and 4 columns: chromosome (chr),
position (pos), reference allele (ref), and alternative allele (alt).

\code{variant_adj_info}: the data frame or matrix of adjusted variant information (unique identifier)
with p_adj rows (listed in the same order as the columns of \code{GTPG_cond}) and 4 columns: chromosome (chr),
position (pos), reference allele (ref), alternative allele (alt), score statistic (U), and variance (V).
}
\description{
The \code{MetaSTAARlite_worker_cov_cond} function takes in genotype, the genotype
of variants to be adjusted for in conditional analysis, the object
from fitting the null model, variant information and adjusted variant information (unique identifier)
to generate the summary statistics file for the given variant-set, adjusting for a given list of variants.
}
\references{
Li, X., et al. (2023). Powerful, scalable and resource-efficient
meta-analysis of rare variant associations in large whole genome sequencing studies.
\emph{Nature Genetics}, \emph{55}(1), 154-164.
(\href{https://doi.org/10.1038/s41588-022-01225-6}{pub})
}
